clients/drcachesim/CMakeLists.txt

```diff
@@ -136,6 +136,15 @@ add_exported_library(drmemtrace_func_view STATIC tools/func_view.cpp)
 configure_DynamoRIO_standalone(drmemtrace_opcode_mix)
 configure_DynamoRIO_standalone(drmemtrace_view)
 
+add_exported_library(drmemtrace_trace_filter STATIC tools/trace_filter.cpp
+  reader/reader.cpp
```

Comment:
On subclass-ing analyzer_t:
- `analyzer_t` works with `reader_t` which returns `memref_t` objects, whereas the trace filter needs to iterate over `trace_entry_t` objects, so it doesn't fit out of the box.
- we don't really need much of the functionality of `analyzer_t`: the support to run multiple tools, serial/parallel mode. One could think of different filters being different tools, but it's much simpler to have them as command-line flags that enable/disable boolean-returning-routines in `trace_filter_t`.
- here, we want to be able to write trace files, which is not typical of analyzers

We could of course change `analyzer_t` to allow us to do all of this, but to  me it seemed simpler to create a standalone tool, with some duplication from `analyzer_t` (some of it could possible be put into a common library I guess), instead of changing `analyzer_t` too much in ways to support two really different paths/use-cases.

Reply:
I guess I don't see future filters fitting into the mold of small functions all inside a single file.  That mold seems to only apply to one of the 5+ filters we've thought about.  Various filters we've proposed include region-of-interest extraction, which would likely take in a config file and have to parse it which seems a little much for a small function; truncating endpoints to eliminate staggered starts and ends, which might be most naturally implemented as two passes; applying cache filters, which might want an embedded cache simulator; creating per-core traces, which might want to operate in serial mode.  These do not seem like simple single-function operations.  Some seem like they want both a high-level memref_t iteration and a trace record editing iteration at the same time, actually, and more than one might want serial vs parallel modes.  While we may not implement any of these ourselves as we have either decided to go with alternatives or they aren't high priority, they show that there are a variety of complex tools in this space: they are not all going to be very simple "delete records of static type X" filters that take just a few lines of code deciding whether to keep or throw out each record.  It may not be worth building a too-fancy infrastructure prematurely, but it seems better to keep things open and flexible with separate components with an eye on the future.

If we don't subclass there should at least be comments acknowledging the code duplication.  It still seems that analyzer_t is not really tied to the particular iterator, and its general patterns of opening readers, managing thread pools, running streaming tools are the same for both use cases and it could be subclassed without much change.  Given that the other future filters might each need different framework shapes (memref + trace_entry at once; multiple outputs; etc.) that are more complex to all handle at once, maybe starting with something more specific for just one filter is reasonable; but let's document that we think it will need generalization and refactoring to work with other filters.

Maybe get a 3rd opinion too: maybe @bete0 could take a look or we could have an offline discussion.

Reply:
I forgot another filter we wanted: stripping encoding entries and function trace entries for confidentiality reasons.  That one is a simple filter.

Reply:
> which seems a little much for a small function

I see the value in passing in a list of `filter_func_t` class objects while constructing `trace_filter_t`, each of which export a bool-returning `should_filter_out(trace_entry_t*)` routine. Taking the filtering logic as an example (which is currently the most complex filter we need), it will certainly make it cleaner.

> truncating endpoints to eliminate staggered starts and ends, which might be most naturally implemented as two passes

Yes, such use cases will likely take two passes. I can imagine that the `filter_func_t` proposed above will then require subclasses to implement an additional `pre_filter_pass()` routine, and `trace_filter_t` will make two loops over the trace.  Multi-pass filters can be left to future work. It's not required currently. 

>  Some seem like they want both a high-level memref_t iteration

trace_filter_t needs to primarily iterate over trace_entry_t. If we do need something in the form of memref_t we can reuse the converter code from reader_t; but sometimes it may not be as simple. I think the current use cases can work with trace_entry_t though.

> one might want serial vs parallel modes.

I couldn't think of any filtering use case that would need serial iteration. My thought was that filters can always treat each thread's trace separately. Do you have any such use case in mind?

> If we don't subclass there should at least be comments acknowledging the code duplication. 

Will do.

>  It still seems that analyzer_t is not really tied to the particular iterator

I agree that there's enough could-be-shared code to want to share more implementation with analyzer_t. I based my decision on the many usages of `reader_t` in the analyzer_t public API  and many in private implementation details too. Also, as I said, the thread shard data needs to be different since we have to also write, and not just read.

Reply:
I was going to do that per-thread, since it's much faster to not serialize, especially for super long traces.

Reply:
Converted analyzer_t to a template in https://github.com/DynamoRIO/dynamorio/pull/5720.